fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -550,51 +556,67 @@ struct PhysicalShard {
 		rocksdb::Status status;
 		if (format == DataMoveRocksCF) {
 			rocksdb::ExportImportFilesMetaData metaData = getMetaData(checkpoint);
-			rocksdb::ImportColumnFamilyOptions importOptions;
-			importOptions.move_files = true;
-			status = db->CreateColumnFamilyWithImport(getCFOptions(), id, importOptions, metaData, &cf);
-
-			if (!status.ok()) {
-				logRocksDBError(status, "RocksImportColumnFamily");
+			if (metaData.files.empty()) {
+				TraceEvent(SevInfo, "RocksDBRestoreEmptyShard")
+				    .detail("Shard", id)
+				    .detail("CheckpointID", checkpoint.checkpointID);
+				status = db->CreateColumnFamily(getCFOptions(), id, &cf);
+			} else {
+				TraceEvent(SevInfo, "RocksDBRestoreCF");
+				rocksdb::ImportColumnFamilyOptions importOptions;
+				importOptions.move_files = SERVER_KNOBS->ROCKSDB_IMPORT_MOVE_FILES;
+				status = db->CreateColumnFamilyWithImport(getCFOptions(), id, importOptions, metaData, &cf);
+				TraceEvent(SevInfo, "RocksDBRestoreCFEnd").detail("Status", status.ToString());
 			}
 		} else if (format == RocksDBKeyValues) {
 			std::vector<std::string> sstFiles;
 			const RocksDBCheckpointKeyValues rcp = getRocksKeyValuesCheckpoint(checkpoint);
 			for (const auto& file : rcp.fetchedFiles) {
-				TraceEvent(SevDebug, "RocksDBRestoreFile")
-				    .detail("Shard", id)
-				    .detail("CheckpointID", checkpoint.checkpointID)
-				    .detail("File", file.toString());
-				sstFiles.push_back(file.path);
+				if (file.path != emptySstFilePath) {
+					sstFiles.push_back(file.path);
+				}
 			}
 
+			TraceEvent(SevDebug, "RocksDBRestoreFiles")
+			    .detail("Shard", id)
+			    .detail("CheckpointID", checkpoint.checkpointID)
+			    .detail("Files", describe(sstFiles));
+
 			if (!sstFiles.empty()) {
 				ASSERT(cf != nullptr);
 				rocksdb::IngestExternalFileOptions ingestOptions;
-				ingestOptions.move_files = true;
-				ingestOptions.write_global_seqno = false;
-				ingestOptions.verify_checksums_before_ingest = true;
+				ingestOptions.move_files = SERVER_KNOBS->ROCKSDB_IMPORT_MOVE_FILES;
+				ingestOptions.verify_checksums_before_ingest = SERVER_KNOBS->ROCKSDB_VERIFY_CHECKSUM_BEFORE_RESTORE;
 				status = db->IngestExternalFile(cf, sstFiles, ingestOptions);
-				if (!status.ok()) {
-					logRocksDBError(status, "RocksIngestExternalFile");
-				}
 			} else {
 				TraceEvent(SevWarn, "RocksDBServeRestoreEmptyRange")
```

Comment:
Is this a valid scenario? Could the CF be nullptr?

Reply:
Do you mean the `sstFiles.empty()` branch? It is possible, i.e., the source physical shard is empty. cf should not be nullptr, I moved the check above.